### PR TITLE
MN-459: Created Slider component

### PR DESCRIPTION
**Pull request by dlymonk**

### CODE REVIEW

File: packages/public/common-ui-web/src/components/Slider/Slider.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import React, { useRef } from 'react';
+import { useInteractiveStatus } from '@monkvision/common';
+import { SliderProps, useSliderStyle } from './useSliderStyle';
+import { useSlider } from './useSlider';
+
+/**
+ * Slider component allows users to select a value within a specified range by dragging a thumb along a horizontal track.
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/Slider.tsx:7]: A Slider component that allows *

et enlever "a thumb"

File: packages/public/common-ui-web/src/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function Slider({
+  min = 0,
+  max = 100,
+  value = (max - min) / 2,
+  primaryColor = 'secondary-xlight',
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/Slider.tsx:13]: For me :
- primaryColor = knob color
- secondaryColor = selected bar color
- tertiaryColor = NOT selected bar color

File: packages/public/common-ui-web/src/components/Slider/Slider.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  secondaryColor = 'primary',
+  tertiaryColor = 'primary',
+  disabled = false,
+  step = 1,
+  onChange,
+  style = {},
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/Slider.tsx:19]: pas de default value (default = undefined)

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
+import React, { useState, useEffect } from 'react';
+
+/**
+ * Computes the initial position of the slider thumb based on the given value within the specified range.
+ * @param min - The minimum value of the range.
+ * @param max - The maximum value of the range.
+ * @param value - The current value of the slider.
+ * @returns A number representing the initial position of the slider thumb, scaled between 0 and 100.
+ */
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:9]: no comment on non-exported stuff

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+function computeFirstThumbPosition(min: number, max: number, value: number) {
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:10]: getFirstThumbPosition

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+/**
+ * Computes the position of the slider thumb based on the user input event, slider dimensions, and configuration parameters.
+ * @param event - The mouse event or touch event triggering the position computation.
+ * @param sliderRef - Reference to the HTMLDivElement representing the slider.
+ * @param step - The step value for the slider.
+ * @param min - The minimum value of the slider range.
+ * @param max - The maximum value of the slider range.
+ * @returns The computed position of the slider thumb as a percentage.
+ */
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:28]: no comment on non-exported stuff

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+function computeThumbPosition(
+  event: MouseEvent | TouchEvent,
+  sliderRef: React.RefObject<HTMLDivElement>,
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:31]: Remove `React.` and `import { RefObject } from 'react';`

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  max: number,
+): number {
+  if (!sliderRef.current) {
+    return 0;
+  }
+
+  const sliderElement = sliderRef.current;
+  const sliderRect = sliderElement.getBoundingClientRect();
+
+  // Compute the x-coordinate relative to the slider element
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:43]: remove comments in code in page

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+/**
+ * Computes the position of the slider thumb based on the user input event, slider dimensions, and configuration parameters.
+ * @param event - The mouse event or touch event triggering the position computation.
+ * @param sliderRef - Reference to the HTMLDivElement representing the slider.
+ * @param step - The step value for the slider.
+ * @param min - The minimum value of the slider range.
+ * @param max - The maximum value of the slider range.
+ * @returns The computed position of the slider thumb as a percentage.
+ */
+function computeThumbPosition(
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:29]: getThumbPosition

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+/**
+ * Computes the new value based on the rounded percentage position of the slider thumb, the slider's range, and step value.
+ * @param roundedPercentage - The rounded percentage position of the slider thumb.
+ * @param max - The maximum value of the slider range.
+ * @param min - The minimum value of the slider range.
+ * @param step - The step value for the slider.
+ * @returns The new value computed based on the slider's configuration.
+ */
+function computeNewValue(roundedPercentage: number, max: number, min: number, step: number) {
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:68]: getNewSliderValue + add return type

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    const nbDigitAfterDot = step.toString().split('.')[1].length;
+    multiplier = 10 ** nbDigitAfterDot;
+  }
+  return Math.round(((max - min) * (roundedPercentage / 100) + min) * multiplier) / multiplier;
+}
+
+/**
+ * Props accepted by the userSlider hook.
+ */
+export interface useSliderProps {
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:80]: UseSliderParams

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  /** Reference to the HTMLDivElement representing the slider. */
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:81]: remove doc on this interface

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+  sliderRef: React.RefObject<HTMLDivElement>;
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:82]: RefObject only

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+   * @param value - The new value of the slider.
+   */
+  onChange?: (value: number) => void;
+}
+
+/**
+ * Custom hook for creating a slider component with customizable behavior.
+ * @param options - The configuration options for the slider component.
+ * @returns An object containing the current thumb position, a handler to start dragging the thumb, and a boolean indicating if the thumb is currently being dragged.
+ */
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:110]: no comment on non-exported stuff

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    sliderRef?.current?.addEventListener('click', handleClick);
+
+    return () => {
+      document.removeEventListener('mouseup', handleEnd);
+      document.removeEventListener('mousemove', handleMove);
+      document.removeEventListener('touchend', handleEnd);
+      document.removeEventListener('touchmove', handleMove);
+      sliderRef?.current?.removeEventListener('click', handleClick);
+    };
+  }, [isDragging]);
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:167]: verify that placing this dependency does not make the useEffect be called too often

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+import { ColorProp, InteractiveStatus } from '@monkvision/types';
+import { useMonkTheme } from '@monkvision/common';
+import { styles } from './Slider.styles';
+
+/**
+ * Props that the Slider component can accept.
+ */
+export interface SliderProps {
+  /**
+   * The minimum value of the slider.
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:11]: empty line between doc and @default (in whole file)

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+  max?: number;
+
+  /**
+   * The current value of the slider.
+   * @default (max - min) / 2
+   */
+  value?: number;
+
+  /**
+   * The primary color used for the background of the slider.
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:29]: Update color documentation based on my previous comment

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+  step?: number;
+
+  /**
+   * Callback function invoked when the slider value changes.
+   * @param value - The new value of the slider.
+   */
+  onChange?: (value: number) => void;
+
+  /**
+   * Optional styling
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:65]: Specify in this doc that the `width` property is used to set the width of the slider but the height does nothing

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+   */
+  style?: CSSProperties;
+}
+
+/**
+ * Parameters required for calculating the styles of the Slider component.
+ */
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:72]: no comment on non-exported stuff

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+type SliderStylesParams = Required<
+  Pick<SliderProps, 'primaryColor' | 'secondaryColor' | 'tertiaryColor' | 'style'>
+> & {
+  status: InteractiveStatus;
+};
+
+/**
+ * Styles generated for the Slider component.
+ */
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:81]: no comment on non-exported stuff

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+export interface SliderStyles {
+  sliderStyle: CSSProperties;
+  trackStyle: CSSProperties;
+  thumbStyle: CSSProperties;
+  valueTrackStyle: CSSProperties;
+  hoverThumbStyle: CSSProperties;
+}
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:88]: remove this type and remove type definition for return value of hook

File: packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+
+/**
+ * Custom hook for generating styles for the Slider component based on provided parameters.
+ * @param params - Parameters for generating Slider styles.
+ * @returns SliderStyles object containing CSS properties for various slider elements.
+ */
+export function useSliderStyle(params: SliderStylesParams): SliderStyles {
+  const { utils } = useMonkTheme();
+  const { primary, secondary, tertiary } = useMemo(
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSliderStyle.ts:97]: remove usememo

File: packages/public/common-ui-web/test/components/Slider.test.tsx [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+    fireEvent.mouseDown(sliderEl);
+    fireEvent.click(sliderEl);
+    fireEvent.mouseMove(sliderEl, { clientX: 100 });
+    fireEvent.mouseUp(sliderEl);
+    expect(onChange).toHaveBeenCalledTimes(2);
+
+    unmount();
+  });
+
+  it('should diplay and change left CSSproperty when mouseMove event is trigeer', () => {
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/test/components/Slider.test.tsx:96]: trigger *

File: packages/public/common-ui-web/test/components/Slider.test.tsx [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+    const onChange = jest.fn();
+    HTMLElement.prototype.getBoundingClientRect = jest.fn(() => ({
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/test/components/Slider.test.tsx:98]: try to replace every `HTMLElement.prototype.getBoundingClientRect` with things like:

```javascript
// Get ref of div
const slider = screen.getByTestId(...);
const spy = jest.spyOn(slider, 'getBoundingClientRect').mockImplementation(() => { ... });

// And then if you need to check if the function was called :
expect(spy).toHaveBeenCalled();
```

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+): number {
+  let multiplier = 1;
+  if (!Number.isInteger(step)) {
+    const nbDigitAfterDot = step.toString().split('.')[1].length;
+    multiplier = 10 ** nbDigitAfterDot;
+  }
+  return Math.round(((max - min) * (roundedPercentage / 100) + min) * multiplier) / multiplier;
+}
+
+export interface useSliderParams {
```
Comment (souyahia-monk) [inline packages/public/common-ui-web/src/components/Slider/useSlider.ts:51]: Capital letter for this interface : `UseSliderParams`

**Comment by souyahia-monk** [comment]
Write documentation in the README file for the common-ui-web package : add a section for the Slider component (note that the sections of the page are ordered alphabetically)

File: packages/public/common-ui-web/src/components/Slider/useSlider.ts (hunk 25 above)
Comment (dlymonk) [reply packages/public/common-ui-web/src/components/Slider/useSlider.ts]: eslint naming convention added to get rid of this error

**Review by souyahia-monk — APPROVED** [review]

**Pull request merged**